doc/mid-based-signalling.md

```diff
@@ -0,0 +1,276 @@
+organised by user ID / device ID / media group UUID, but they could also just be a flat
+list of tracks. It would be more duplication but maybe less effort to read.
+
+The expected behaviour here would be for foci to essentially maintain a structure with all
+tracks being pushed to it. This structure would probably have call IDs as a top level index,
+then look very similar to the structure of the `m.call.advertise` event. It could either keep
+a reference to the transceiver it was receiving media on in the structure itself alongside
+the media UUID, or maintain a separate map of media UUID to transceiver / peer connection such
+that the first structure could be marshalled to JSON and sent to clients as-is.
+
+On the client side, the client will essentially take the `m.call.advertise` data and save it
+almost as-is. It would probably cross-reference it against the call member state events to
+ensure that it wasn't showing feeds for any users that did not have state events indicating
+that they were in the call, although if we trust the focus and assume that conf IDs are unique
+enough to be unguessable, this may be unnecessary.
+
+In the simplest case, the client will simply iterate through this structure and add every
+media UUID it finds to an `m.call.subscribe` message.
+
+The most complex part is that when the `m.call.describe` message arrives back from the focus,
+it will have to search through the data from the `m.call.advertise` message to map the tracks
+it is now receving to the right user IDs (XXX: it could build them into a map to make this lookup
+efficient, although if we make the advertise message indexed by media UUID then it already has
+a map indexed by the correct thing...)
+
+The `select_answer` is also tweaked to be more extensible-event like although is essentially
+the same:
+
+m.select\_answer
+```
+"m.select_answer": {
+    "version": "1",
+    "conf_id": "1674732106391mz4ygIc84Q2Z6mJ5",
+    "call_id": "35657a5b793ce",
+    "party_id": "123456",
+    "selected_party_id": "678910",
+}
+```
+
+Once the client receives this, it decides what tracks it wants to receive and then sends
+a subscribe message over the data channel:
+
+m.call.subscribe
+```
+"m.call.subscribe": {
+    "seq": 1,
+    "media_uuids": {
+        "aaaa-aaaa-aaaaaa-aaaa-aaaa": {
+            "width": 1024,
+            "height": 576,
+        },
+        "bbbb-bbbb-bbbbbb-bbbb-bbbb": {},
+    },
+},
+```
+
+This has also been rearrnaged a little to make the media UUIDs the keys and remove the
+unsubscribe section which is unnecessary if we always send the complete set of tracks we
+want to receive (we unsubscribe by just removing the media UUID from the dict).
+
+This also now contains a sequence number. This is a monotonically increasing integer, starting
+at 0 and scoped to the lifetime of the peer connection. The focus will send a reply containing
+this sequence number to acknowledge that it has processed the message. This can be a positive ack:
```

Comment:
Nice 🚀 

Btw, do we want to use a sequence number or rather some sort of a randomly generated transaction ID?

I know that the advantage of a sequence number is that it still can be used as a transaction ID and we probably need it since To-Device messages are not ordered. The only problem that I can think of though is that what if one party receives a message with a sequence number `10000` after a message with a sequence number `5`? What's the behavior? Should we drop the `10000` message or should we buffer it and process it once the previous message arrives (this opens up a possible attack vector of sending 1, 2, 3, 10000000 and causing the receiving client to deplete their own buffer)?

Also, what to do when the SFU gets restarted and does not know the last sent `seq`? - I suppose this is solved by "scoped to the lifetime of the peer connection", but I wonder if we need to have `seq` for all events (negotiate events could also fail).

Reply:
True, seqnums allow for a little more flexibility but we wouldn't need the re-ordering capabilities if we're using a channel that guarantees ordering which the WebRTC DC will give us. A randomly generated ID could avoid clients thinking they have to do reordering and they wouldn't have to worry about masking them monotonically increasing.